### PR TITLE
RF: move load/save logic out of SpatialImage

### PR DESCRIPTION
**Pull request by bcipolli**
This PR precedes #360, which builds off of this work. I'll copy some text from that PR; there's quite a bit of discussion about how this fits into forward-looking design of surface vs. spatial, binary vs. xml object models.

**Issue:**
The final issue blocking streamlined GIFTI/CIFTI in nibabel is: image file I/O functions are tied directly to the `SpatialImage` class; non-volumetric images (GIFTI/CIFTI) can't extend `SpatialImage` and so can't easily get in the `load`/`save` patheway.

**Proposed Solution:**
The good news is, the file I/O code actually is quite general. We can easily abstract the file I/O into a higher-level class (which I've called `FileBasedHeader`/`FileBasedImage`, as the more generic `Header` was already taken) that `SpatialImage` and `GiftiImage` inherit from.

**To do:**
- [X] Move file I/O from `SpatialImage` into `FileBasedHeader`/`FileBasedImage` 
- [X] Update tests that assume all image types are of type `SpatialImage`

**Notes:**
* Given the above design, I believe this is mostly done. Rather than fine-tuning the code now, I'd rather get ideas about the overall design first. @matthew-brett and @effigies , your input is most welcome when you have the time!
* Whatever we decide here, I'll be using to push forward GIFTI (#360) and CIFTI (#249) support

### CODE REVIEW

File: nibabel/spatialimages.py [hunk 1]
```diff
@@ -312,24 +308,18 @@ def supported_np_types(obj):
     return set(supported)
 
 
-class ImageDataError(Exception):
-    pass
+class Header(SpatialHeader):
+  '''Alias for SpatialHeader; kept for backwards compatibility.'''
```
Comment (effigies) [inline nibabel/spatialimages.py]: Should this be marked deprecated? e.g.

    def __init__(self, *args, **kwargs):
        warnings.warn('Header is deprecated, use SpatialHeader', DeprecationWarning)
        SpatialHeader.__init__(self, *args, **kwargs)

File: nibabel/filebasedimages.py [hunk 2]
```diff
@@ -0,0 +1,569 @@
+# emacs: -*- mode: python-mode; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the NiBabel package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+''' Simple interfaces for images'''
```
Comment (effigies) [inline nibabel/filebasedimages.py]: Maybe 'Common interface for binary and XML image formats'?

File: nibabel/filebasedimages.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+
+try:
+    basestring
+except NameError:  # python 3
+    basestring = str
```
Comment (effigies) [inline nibabel/filebasedimages.py]: from .externals.six import string_types

And replace `basestring` with `string_types` throughout.

File: nibabel/filebasedimages.py [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+       >>> # read it back again from the written files
+       >>> img2 = nib.AnalyzeImage.from_file_map(file_map)
+       >>> np.all(img2.get_data() == data)
+       True
+       >>> # write, read it again
+       >>> img2.to_file_map()
+       >>> img3 = nib.AnalyzeImage.from_file_map(file_map)
+       >>> np.all(img3.get_data() == data)
+       True
+    '''
```
Comment (effigies) [inline nibabel/filebasedimages.py]: Much of these docs seem to be specific to `SpatialImage`.

File: nibabel/filebasedimages.py [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+        return self._header
+
+    def __getitem__(self):
+        ''' No slicing or dictionary interface for images
+        '''
+        raise TypeError("Cannot slice image objects; consider slicing image "
+                        "array data with `img.dataobj[slice]` or "
+                        "`img.get_data()[slice]`")
+
+    def get_header(self):
```
Comment (effigies) [inline nibabel/filebasedimages.py]: Mark deprecated?

File: nibabel/filebasedimages.py [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+        """ Get header from image
+
+        Please use the `header` property instead of `get_header`; we will
+        deprecate this method in future versions of nibabel.
+        """
+        return self.header
+
+    def get_filename(self):
```
Comment (effigies) [inline nibabel/filebasedimages.py]: Worth propertizing `filename`? Or keep this?

File: nibabel/filebasedimages.py [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+
+        Returns
+        -------
+        cimg : ``spatialimage`` instance
+           Image, of our own class
+        '''
+        return klass(img.dataobj,
+                     img.affine,
+                     klass.header_class.from_header(img.header),
+                     extra=img.extra.copy())
```
Comment (effigies) [inline nibabel/filebasedimages.py]: Does this work for non-`SpatialImage`s?

File: nibabel/spatialimages.py [hunk 8]
```diff
@@ -136,15 +136,11 @@
     basestring
 except NameError:  # python 3
     basestring = str
```
Comment (effigies) [inline nibabel/spatialimages.py]: `basestring` unused in this file.

File: nibabel/spatialimages.py [hunk 9]
```diff
@@ -312,24 +308,18 @@ def supported_np_types(obj):
     return set(supported)
 
 
-class ImageDataError(Exception):
-    pass
+class Header(SpatialHeader):
+  '''Alias for SpatialHeader; kept for backwards compatibility.'''
+  pass
```
Comment (effigies) [inline nibabel/spatialimages.py]: 4-space indentation.

**Comment by bcipolli** [comment]
Thanks @effigies! 

A question: for some comments, it's about old code. Most of the code is copy-pasted from `spatialimage.py`; it seemed dangerous to move a bunch of code, and make changes at the same time (as git doesn't really track those changes as different).

I'm happy to deal with those things here, but I wanted to double-check first. Thanks!

**Comment by effigies** [comment]
Could be a separate PR, but my thought is that while people are looking at code is a good time to make changes. And the changes will be tracked in the commits, if not the full PR diff.

File: nibabel/spatialimages.py (hunk 1 above)
Comment (bcipolli) [inline nibabel/spatialimages.py]: :+1:

File: nibabel/filebasedimages.py (hunk 2 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: I put: `''' Common interface for any image format--volume or surface, binary or xml.'''`

File: nibabel/spatialimages.py (hunk 1 above)
Comment (effigies) [inline nibabel/spatialimages.py]: I should probably note that you'll want `stacklevel=2` in that warning.

File: nibabel/filebasedimages.py (hunk 3 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: :+1:

File: nibabel/filebasedimages.py (hunk 4 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: Moved this to the `SpatialImage` class docstring. Also fixed up the docs; they were a bit hacked up.

File: nibabel/filebasedimages.py (hunk 5 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: :+1:

Comment (bcipolli) [inline nibabel/filebasedimages.py]: There are a number of uses of `get_header` in the code that I'll change over to `header` as well.

Comment (effigies) [inline nibabel/filebasedimages.py]: Think it's just ecat? The `test_image_api` bit should stay until `get_header` is removed entirely.

File: nibabel/filebasedimages.py (hunk 6 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: We can propertize it and document that we'll deprecate it in the future.

Comment (effigies) [inline nibabel/filebasedimages.py]: This one I want @matthew-brett's comment on. Since other get/sets in the vicinity seem to have been deprecated, there might be a good reason this one hasn't been.

File: nibabel/filebasedimages.py (hunk 7 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: No. Anything with `dataobj` or `affine` should not appear in `filebasedimage`. There are also some dirty docs; fixing this now.

File: nibabel/spatialimages.py (hunk 8 above)
Comment (bcipolli) [inline nibabel/spatialimages.py]: :+1:

File: nibabel/filebasedimages.py [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the NiBabel package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+''' Common interface for any image format--volume or surface, binary or xml.'''
+
+from ..externals.six import string_types
```
Comment (effigies) [inline nibabel/filebasedimages.py]: `.externals`, not `..externals`

Comment (bcipolli) [inline nibabel/filebasedimages.py]: Yep. I pushed before running tests; give me a min..

**Comment by bcipolli** [comment]
OK @effigies ; cleaned up and ran tests; pls take a look!

**Comment by bcipolli** [comment]
doctests are failing; will work on those now.

File: nibabel/filebasedimages.py (hunk 6 above)
Comment (matthew-brett) [inline nibabel/filebasedimages.py]: Sorry - might be moot now - but `get_filename` is different because it returns the canonical filename for the image - for example will return `image.img` when input filename might have been `image.hdr`, for a Nifti Pair.

Comment (effigies) [inline nibabel/filebasedimages.py]: For clarity, you think it should stay as `{get,set}_filename`? Or does the entire `filename` interface seem irrelevant, now?

Comment (matthew-brett) [inline nibabel/filebasedimages.py]: On Wed, Oct 21, 2015 at 4:01 PM, Chris Markiewicz <notifications@github.com>
wrote:

> In nibabel/filebasedimages.py
> <https://github.com/nipy/nibabel/pull/364#discussion_r42674196>:
>
> > +    def __getitem__(self):
> > +        ''' No slicing or dictionary interface for images
> > +        '''
> > +        raise TypeError("Cannot slice image objects; consider slicing image "
> > +                        "array data with `img.dataobj[slice]` or "
> > +                        "`img.get_data()[slice]`")
> > +
> > +    def get_header(self):
> > +        """ Get header from image
> > +
> > +        Please use the `header` property instead of `get_header`; we will
> > +        deprecate this method in future versions of nibabel.
> > +        """
> > +        return self.header
> > +
> > +    def get_filename(self):
>
> For clarity, you think it should stay as {get,set}_filename? Or does the
> entire filename interface seem irrelevant, now?
>

Sorry not to be clear - I think, while the FileBasedImage can have more
than one filename used to identify the file (e.g '.img', '.hdr') it should
stay as `get/set_filename`...

**Comment by bcipolli** [comment]
Made the code updates. A bit worried about deprecating `Header` now; should we simply document that they're equivalent and it will be deprecated (a la `get_header()`)?

Also a bit concerned about changing the base class of images from `Header` to `SpatialHeader`. Would this break `pickle`d headers?

**Comment by effigies** [comment]
@bcipolli Regarding 9cc27ea: For images where there's a `filename` interface (basically, if there CAN be multiple files, then return the canonical file), then `{get,set}_` are [preferred](https://github.com/nipy/nibabel/pull/364#discussion_r42674598). So that commit can be stripped out.

Is there any possibility that XML-based datasets will have multiple files (so a canonical filename is needed)? If not, then `{get,set}_filename` could be moved back into `SpatialImage`. But I don't think there's anything wrong with leaving them in `FileBasedImage`, either.

**Comment by effigies** [comment]
> A bit worried about deprecating `Header` now; should we simply document that they're equivalent and it will be deprecated (a la `get_header`())?

To be honest, I'm not quite sure what the difference is between posting an intent to deprecate and actually deprecating but not removing. Documentation should be updated ASAP, and `DeprecationWarning`s are silent by default.

> Also a bit concerned about changing the base class of images from `Header` to `SpatialHeader`. Would this break pickled headers?

Good question. Is pickling really a thing we care about? The only case I can see is in `tests/test_analyze.py`. I guess I'm confused why you'd pickle an image rather than save the image. But we can try pickling under `Header` and loading under `SpatialHeader`, to be safe.

In any event, I'm also okay with leaving things as subclassed from `Header` and not `SpatialHeader`.

**Comment by bcipolli** [comment]
>that commit can be stripped out.

Done

>Is there any possibility that XML-based datasets will have multiple files

I don't see any reason to keep it off of `FileBasedImage`; there's nothing to suggest surface images couldn't use a similar approach at some point. CIFTI does something close (using multiple sub-extensions to specify different data types related to a study), but not quite.

>Is pickling really a thing we care about?

I'm not sure; perhaps some people or tools use `pickle` or `shelve` to store working sessions.

>In any event, I'm also okay with leaving things as subclassed from Header and not SpatialHeader.

I'd prefer not to do that; I think that introduces complexity without much of the benefits. I suggest we either use `SpatialHeader` and deprecate `Header`, or just have `Header` without `SpatialHeader.

I'll try this out.

**Comment by bcipolli** [comment]
@effigies I removed the commit and tested `shelve` (which uses `pickle`) manually (create a `Minc1Header` in `master`, load up in this branch. No issue.

**Comment by effigies** [comment]
Okay. So no barrier to deprecating `Header` for `SpatialHeader`?

**Comment by bcipolli** [comment]
Not that I see. I agree that deprecating directly is a fine user experience, from my POV

File: nibabel/filebasedimages.py [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+    contain enough information so that an existing image instance can save
+    itself back to the files pointed to in ``file_map``.  When a file holder
+    holds active file-like objects, then these may be affected by the
+    initial file read; in this case, the contains file-like objects need to
+    carry the position at which a write (with ``to_files``) should place the
+    data.  The ``file_map`` contents should therefore be such, that this will
+    work:
+    '''
+    files_types = (('image', None),)
+    alternate_exts = ()  # Modified by @ImageOpener.register_ext_from_image
```
Comment (effigies) [inline nibabel/filebasedimages.py]: `valid_exts`

File: nibabel/filebasedimages.py [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+        -------
+        cimg : ``spatialimage`` instance
+           Image, of our own class
+        '''
+        raise NotImplementedError()
+
+    @classmethod
+    def is_valid_extension(klass, ext):
+        valid = tuple(ft[1] for ft in klass.files_types) + klass.alternate_exts
+        return ext.lower() in valid
```
Comment (effigies) [inline nibabel/filebasedimages.py]: `is_valid_extension` was removed from `SpatialImage` in 0c06464.

File: nibabel/filebasedimages.py (hunk 11 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: `_meta_sniff_len` is optional, and on the header.

`valid_exts` is required, but is also required to be defined by each subclass. Defining it here makes it easier to forget to do that, so I excluded it.

Comment (effigies) [inline nibabel/filebasedimages.py]: `sizeof_hdr` is on the header, [`_meta_sniff_len`](https://github.com/nipy/nibabel/pull/364/files#diff-1053478051df4b99b0c69a5365f5f0beL326) is on the image and used in [`path_maybe_image`](https://github.com/nipy/nibabel/pull/364/files#diff-496e321b697d515cfea0150d3c97dfb5R530) without being checked to exist, first. (It does check whether the header class defines `may_contain_header`, but that's not tied to presence of `_meta_sniff_len`, AFAICT.)

`files_types` must also be defined by subclasses, correct? I think its purpose here is documentary. But the point of my comment was that `alternate_exts` has been removed from `SpatialImage`.

**Comment by effigies** [comment]
Had another read through, and once we agree on the above issues, I think this can go in.

File: nibabel/filebasedimages.py (hunk 11 above)
Comment (bcipolli) [inline nibabel/filebasedimages.py]: :+1:

**Comment by bcipolli** [comment]
@effigies Made the tweaks requested, tests pass... I think we're good to go! Excited! This is the last refactoring needed to really jump into #249 for CIFTI :)

**Comment by effigies** [comment]
Pushed a PEP8 cleanup for `filebasedimages.py`, as that was easier than pointing them out. Have a suspicion removing the `numpy` import will break the doctests. If so, I'll put it into the doctests themselves.

**Comment by bcipolli** [comment]
>Pushed a PEP8 cleanup for filebasedimages.py, as that was easier than pointing them out.

Agree, thanks for doing that!

**Comment by effigies** [comment]
Alright. If you're happy with that commit (and the tests pass), in it goes.

**Comment by bcipolli** [comment]
:+1:
